cip/1.accepted/CIP2017-02-07-Map-Projection.adoc

```diff
@@ -0,0 +1,121 @@
+
+Note that while it is typically superfluous to use both an _all properties selector_ and explicit _property selectors_, this is syntactically allowed.
+Sometimes this might be the most convenient way to express the desired projection, since the explicit _property selector_ allows further projection of the nested property, and this value takes precedence over the _all properties selector_.
+
+=== Syntax
+
+The EBNF syntax of map projection takes the following place in the Cypher syntax:
+
+[source, ebnf]
+----
+Atom = … | MapProjection | … ;
+
+MapProjection = Variable, ProjectedMap ;
+
+ProjectedMap = '{', (AllPropertiesSelector | ProjectionEntry), {',', ProjectionEntry}, '}' ;
+
+ProjectionEntry = LiteralEntry
+                | PropertySelector
+                | VariableSelector
+                ;
+LiteralEntry = PropertyKeyName, ':', Expression ;
+PropertySelector = '.', PropertyKeyName, [ProjectedMap] ;
+VariableSelector = Variable, [ProjectedMap] ;
+AllPropertiesSelector = '*', {'-', PropertyKeyName} ;
+
+ReturnItem = Expression
+           | (Expression, 'AS', Variable)
+           | (ProjectedMap, 'AS', Variable)
+           ;
+----
+
+=== Examples
+
+[source, cypher]
+.Fetch name and address for a person
+----
+MATCH (person:Person{userId=$user})-[:ADDRESS]->(address)
+RETURN person {.firstName, .lastName, id: $user,
+               address {.streetAddress, .city, .postalCode} }
+----
+
+.Example result
+----
++----------------------------------------------------------------------------------------+
+| person                                                                                 |
++----------------------------------------------------------------------------------------+
+| {firstName: "Sherlock", lastName: "Holmes", id: "0099CC",                              |
+|  address: {streetAddress: "221B Baker Street", city: "London", postalCode: "NW1 6XE"}} |
++----------------------------------------------------------------------------------------+
+----
+
+[source, cypher]
+.Bind variables into a map
+----
+MATCH (group:Group{name:$groupName})<-[:MEMBER_OF]-(member)-[:KNOWS]-(friend)-[:ADDRESS]->(address)
+RETURN member.name, collect( {friend{.name},address{.city}} ) AS friends
+----
+
+.Example result
+----
++-------------+---------------------------------------------------------------+
+| member.name | friends                                                       |
++-------------+---------------------------------------------------------------+
+| John Doe    | • {friend: {name:"Heather Taylor"}, address:{city:"Miami"}}   |
+|             | • {friend: {name:"Leroy Jenkins"}, address:{city:"New York"}} |
++-------------+---------------------------------------------------------------+
+| Eve Longman | • {friend: {name:"Mickey Mouse"}, address:{city:"Orlando"}}   |
+|             | • {friend: {name:"Leroy Jenkins"}, address:{city:"New York"}} |
+|             | • {friend: {name:"Minnie Mouse"}, address:{city:"Orlando"}}   |
```

Comment:
Cypher style: single tick for literal string.
Cypher style: space after colons.